packages/sources/uniswap-v3/src/endpoint/crypto.ts

```diff
@@ -0,0 +1,136 @@
+
+  const outputAmount = new Decimal(output.toString()).div(new Decimal(10).pow(toDecimals))
+  const rate = outputAmount.div(inputAmount)
+
+  const data: ResponseSchema = {
+    input: amount.toString(),
+    inputToken: from,
+    inputDecimals: fromDecimals,
+    output: output.toString(),
+    outputToken: to,
+    outputDecimals: toDecimals,
+    rate: rate.toNumber(),
+  }
+
+  const response = {
+    status: 200,
+    statusText: 'OK',
+    headers: {},
+    config: {},
+    data: data,
+  }
+  const result = Requester.validateResultNumber(response.data, [resultPath])
+
+  return Requester.success(jobRunID, Requester.withResult(response, result), config.verbose)
+}
+
+/**
+ * getTokenDetails will find the address and number of decimal for a token.
+ *
+ * The order of operations is as follows:
+ *  - address:
+ *     1. Check if the address was provided in the request.
+ *     2. If not, check the symbol in the request to see if we have pre-set the address for this symbol/network.
+ *     3. If not, we assume the symbol param was actually the address.
+ *  - decimals:
+ *     1. Check if the number of decimals was provided in the request.
+ *     2. Query the contract at the address found above to see how many decimals it's set to.
+ * @param validator The validation class to use
+ * @param direction Used to get the params in the request
+ * - `{direction}` is the symbol of the token (to find pre-set token details)
+ * - `{direction}Address` is the token address as set in the request
+ * - `{direction}Decimals` is the number of decimals for the token as set in the request
+ * @param config Configuration to extract token decimals from
+ */
+const getTokenDetails = async (
+  validator: Validator,
+  direction: string,
+  config: Config,
+): Promise<{ address: string; decimals: number }> => {
+  const symbol = validator.overrideSymbol(
+    AdapterName,
+    validator.validated.data[direction],
+  ) as string
+  const address =
+    validator.validated.data[`${direction}Address`] ||
+    validator.overrideToken(symbol, config.network) ||
+    symbol
+  const decimals =
+    validator.validated.data[`${direction}Decimals`] || (await getDecimals(address, config))
+
+  return { address, decimals }
+}
+
+const getDecimals = async (address: string, config: Config): Promise<number> =>
+  new ethers.Contract(address, erc20ABI, config.provider).decimals()
+
+const getBestRate = async (
+  from: string,
+  to: string,
+  amount: BigNumber,
+  config: Config,
+): Promise<[output: BigNumber]> => {
+  // pull abi from file
+  const quoterContract = new ethers.Contract(config.uniswapQuoter, quoterABI, config.provider)
+
+  // encode inputs for quoteExactInputSingle
+  const sqrtPriceLimitX96 = 0
+
+  // execute non view function as a call
+  const result = await quoterContract.callStatic.quoteExactInputSingle(
+    from,
+    to,
+    config.feeAmount,
+    amount,
+    sqrtPriceLimitX96,
+  )
+
+  return [result]
```

Comment:
Since we only have a single value, can we just return that instead?

```suggestion
  return await quoterContract.callStatic.quoteExactInputSingle(
    from,
    to,
    config.feeAmount,
    amount,
    sqrtPriceLimitX96,
  )
```